cupy/core/_routines_manipulation.pyx

```diff
@@ -559,18 +559,31 @@ cpdef ndarray concatenate_method(tup, int axis):
     if not have_same_types:
         dtype = functools.reduce(numpy.promote_types,
                                  set([a.dtype for a in arrays]))
-    return _concatenate(arrays, axis, tuple(shape), dtype)
+    if out is None:
+        out = ndarray(shape, dtype=dtype)
+    else:
+        if len(out.shape) != len(shape):
+            raise ValueError('Output array has wrong dimensionality')
+        if out.shape != shape:
+            raise ValueError('Output array is the wrong shape')
+        if out.dtype.kind != dtype.kind:
+            raise TypeError('Cannot cast scalar from dtype(\'{}\')'
+                            ' to dtype(\'{}\') according to the'
+                            ' rule \'same_kind\''.format(dtype, out.dtype))
+
+    return _concatenate(arrays, axis, tuple(shape), dtype, out)
 
 
-cpdef ndarray _concatenate(list arrays, Py_ssize_t axis, tuple shape, dtype):
-    cdef ndarray a, ret
+cpdef ndarray _concatenate(
+        list arrays, Py_ssize_t axis, tuple shape, dtype, ndarray out):
+    cdef ndarray a
     cdef Py_ssize_t i, aw, itemsize, axis_size
     cdef bint all_same_type, same_shape_and_contiguous
     # If arrays are large, Issuing each copy method is efficient.
     cdef Py_ssize_t threshold_size = 2 * 1024 * 1024
 
     if len(arrays) > 8:
-        all_same_type = True
+        all_same_type = out.dtype == arrays[0].dtype
```

Comment:
Don't you need to check all the arrays, instead of just the first one?

Reply:
i compare the out with the first one to verify the out and the first array share the same type, otherwise that flag will be set to false and the slow path is executed.
The comparison with the rest of arrays is done in the for loop below.
I can move the comparison of the out array type after this loop, maybe thats more readable?

Reply:
Ah, I didn't check the code below.

I think what to be done is to remove `dtype` argument from `_concatenate()`, use `dtype = out.dtype` and keep this line intact?

Reply:
I like that approach!

Reply:
I like that approach!